### PR TITLE
WIP: input pipeline refactor

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

File: finetune/base.py [hunk 1]
```diff
@@ -181,27 +182,22 @@ def _n_steps(self, n_examples, batch_size, n_gpus):
...
-            and len(Xs) != len(Y)
-        ):
-            raise FinetuneError(
-                "Mismatch between number of examples ({}) and number of targets ({}) provided.".format(
-                    len(Xs), len(Y)
-                )
+    def finetune(self, Xs, Y=None, context=None, update_hook=None):
+        if callable(Xs):
+            datasets = self.input_pipeline.get_dataset_from_generator(
+                generator_fn, input_mode=InputMode.TRAIN, update_hook=update_hook
```
Comment (madisonmay) [inline finetune/base.py]: I think `generator_fn` needs to be `Xs` yeah?

Comment (madisonmay) [reply finetune/base.py]: If so -- can we add a test for the generator behavior?

File: finetune/base_models/gpt2/featurizer.py [hunk 2]
```diff
@@ -204,9 +204,10 @@ def gpt2_featurizer(
         pool_idx = tf.cast(
             tf.argmax(tf.cast(tf.equal(X[:, :, 0], clf_token), tf.float32), 1), tf.int32
         )
+        batch, length, _ = shape_list(X)
         clf_h = tf.gather(
             clf_h,
-            tf.range(shape_list(X)[0], dtype=tf.int32) * config.max_length + pool_idx,
+            tf.range(batch, dtype=tf.int32) * length + pool_idx,
```
Comment (madisonmay) [inline finetune/base_models/gpt2/featurizer.py:210]: Oh yikes -- was this straight up broken previously? Or were we updating config.max_length all the time or something.

File: finetune/base.py [hunk 3]
```diff
@@ -775,7 +773,7 @@ def finetune_grid_search_cv(
 
     def process_long_sequence(self, X, context=None, **kwargs):
         arr_encoded = [
-            self.input_pipeline._text_to_ids(x) for x in self.input_pipeline._format_for_inference(X)
+            self.input_pipeline._text_to_ids(x) for x in X
```
Comment (lhz1029) [inline finetune/base.py]: Why is this change possible now?

File: finetune/encoding/target_encoders.py [hunk 4]
```diff
@@ -193,7 +193,7 @@ def __init__(self, pad_token="<PAD>"):
         self.lookup = None
 
     def fit(self, labels):
-        self.classes_ = list(set(lab["label"] for lab in labels) | {self.pad_token})
+        self.classes_ = list(set(lab_i["label"] for lab in labels for lab_i in lab) | {self.pad_token})
```
Comment (lhz1029) [inline finetune/encoding/target_encoders.py:196]: Why is this being flattened another layer?

File: finetune/input_pipeline.py [hunk 5]
```diff
@@ -161,274 +137,193 @@ def _compute_class_counts(self, encoded_dataset):
...
+            return {
+                "predict_dataset": batch_dataset(
+		    raw_dataset,
+                    batch_size=self.config.predict_batch_size,
+                    shapes=shapes,
+                )
+            }
+        
+        if self.config.chunk_long_sequences:
+            LOGGER.warning("The dataset size is not adjusted for chunk long sequences when training from a generator")
```
Comment (lhz1029) [inline finetune/input_pipeline.py:192]: It seems like _inference() always calls get_dataset_from_generator. Seems like we might have problems with chunk_long_sequences during prediction then?

**Review by lhz1029 — APPROVED** [review]
Thanks for doing this! Things look great!

File: finetune/encoding/target_encoders.py (hunk 4 above)
Comment (benleetownsend) [reply finetune/encoding/target_encoders.py:196]: Previously the flattening was done by a sequence labellers post_data_initialisation method. this change gives all the label encoders the same interface.

File: finetune/input_pipeline.py (hunk 5 above)
Comment (benleetownsend) [reply finetune/input_pipeline.py:192]: I think the only problem might be the dataset length, but I think this is covered by ProgessBar's new chunked_length argument.

File: finetune/base.py (hunk 3 above)
Comment (benleetownsend) [reply finetune/base.py]: Format for inference was originally setup to cover the case where the inputs were different at this point for training and inference, which was the original sequence labeler before indico_to_finetune was removed. After this, format_for_inference was just a noop

File: finetune/target_models/classifier.py [hunk 6]
```diff
@@ -68,20 +60,22 @@ def featurize(self, X, **kwargs):
...
         :param X: list or array of text to embed.
         :returns: list of class labels.
 
-
         Chunk idx for prediction.  Dividers at `step_size` increments.
         [  1  |  1  |  2  |  3  |  3  ]
         """
+        return super().predict(X, context=context, **kwargs)
+
+    def _predict_internal(self, zipped_data, probas=False, **kwargs):
```
Comment (madisonmay) [inline finetune/target_models/classifier.py]: I don't understand the distinction between `_predict` and `_predict_internal`.  From what I see so far `_predict` just calls `_predict_internal`?

File: finetune/scheduler.py [hunk 7]
```diff
@@ -83,6 +83,6 @@ def featurize(self, model_file, x, *args, **kwargs):
 
     def featurize_sequence(self, model_file, x, *args, **kwargs):
         model = self._rotate_in_model(model_file)
-        seq_features = model.featurize(x, *args, **kwargs)
+        seq_features = model.featurize_sequence(x, *args, **kwargs)
```
Comment (madisonmay) [inline finetune/scheduler.py:86]: Yup thank you.  Copy paste error on my part.

File: finetune/target_models/comparison_regressor.py [hunk 8]
```diff
@@ -60,12 +61,13 @@ def predict(self, pairs, context=None, **kwargs):
         """
         Produces a floating point prediction determined by the fine-tuned model.
 
-
         :param pairs: Array of text, shape [batch, 2]
         :returns: list of floats, shape [batch]
         """
-        return super().predict(pairs, context=context, **kwargs)
-
+        zipped_data = self.input_pipeline.zip_list_to_dict(X=pairs, context=context)
```
Comment (madisonmay) [inline finetune/target_models/comparison_regressor.py:67]: I believe this should be the same as normal predict() now?

File: tests/test_sort_by_length.py [hunk 9]
```diff
@@ -11,42 +11,48 @@ class TestSortByLength(unittest.TestCase):
 
     def test_sort_by_length(self):
         # Construct a fake dataset where lengths vary wildly
-        fake_data = ["A"] * 1000 + ["B " * 120] * 48
+        fake_data = (["A"] * 15 + ["B " * 120]) * 64
 
         # Blend in the long examples
-        random.shuffle(fake_data)
+#        random.shuffle(fake_data)
```
Comment (madisonmay) [inline tests/test_sort_by_length.py]: Can just rm this line

**Review by madisonmay — CHANGES_REQUESTED** [review]
Looks good -- only question is around the _internal_predict fns

**Review by madisonmay — APPROVED** [review]

File: finetune/base.py [hunk 10]
```diff
@@ -508,36 +522,36 @@ def generate_text(self, seed_text="", max_length=None, use_extra_toks=None):
         """
         if use_extra_toks is None:
             use_extra_toks = self._trained
-    
+
```
Comment (lhz1029) [inline finetune/base.py]: we probably want to get rid of all of this extra whitespace (other examples as well)

File: finetune/encoding/target_encoders.py (hunk 4 above)
Comment (lhz1029) [reply finetune/encoding/target_encoders.py:196]: :+1:

File: finetune/target_models/multiple_choice.py [hunk 11]
```diff
@@ -210,4 +213,4 @@ def featurize(self, questions, answers, **kwargs):
         :param answers: List or array of text, shape [n_answers, batch]
         :returns: np.array of features of shape (n_examples, embedding_size).
         """
-        return BaseModel.featurize(self, zip(questions, answers), **kwargs)
+        return super.featurize(self, zip(questions, answers), **kwargs)
```
Comment (lhz1029) [inline finetune/target_models/multiple_choice.py]: Don't we want `super()`?

**Review by lhz1029 — APPROVED** [review]

**Pull request merged**